docarray/typing/url/video_url.py

```diff
@@ -0,0 +1,153 @@
+            False, else only VideoNdArray representing the keyframes.
+
+
+        EXAMPLE USAGE
+
+        .. code-block:: python
+
+            from typing import Optional
+
+            from docarray import BaseDocument
+
+            from docarray.typing import VideoUrl, VideoNdArray, AudioNdArray, NdArray
+
+
+            class MyDoc(BaseDocument):
+                video_url: VideoUrl
+                video: Optional[VideoNdArray]
+                audio: Optional[AudioNdArray]
+                key_frame_indices: Optional[NdArray]
+
+
+            doc = MyDoc(
+                video_url='https://github.com/docarray/docarray/tree/feat-add-video-v2/tests/toydata/mov_bbb.mp4?raw=true'
+            )
+            doc.audio, doc.video, doc.key_frame_indices = doc.video_url.load()
+
+            assert isinstance(doc.video, VideoNdArray)
+            assert isinstance(doc.audio, AudioNdArray)
+            assert isinstance(doc.key_frame_indices, NdArray)
+
+        You can load only the key frames:
+
+        .. code-block:: python
+
+            from typing import Optional
+
+            from docarray import BaseDocument
+
+            from docarray.typing import VideoUrl, VideoNdArray
+
+
+            class MyDoc(BaseDocument):
+                video_url: VideoUrl
+                video_key_frames: Optional[VideoNdArray]
+
+
+            doc = MyDoc(
+                video_url='https://github.com/docarray/docarray/tree/feat-add-video-v2/tests/toydata/mov_bbb.mp4?raw=true'
+            )
+            doc.video_key_frames = doc.video_url.load(only_keyframes=True)
+
+            assert isinstance(doc.video_key_frames, VideoNdArray)
+
+        """
+        import av
+
+        with av.open(self, **kwargs) as container:
+            if only_keyframes:
+                stream = container.streams.video[0]
+                stream.codec_context.skip_frame = 'NONKEY'
+
+            audio_frames = []
+            video_frames = []
+            keyframe_indices = []
+
+            for frame in container.decode():
+                if type(frame) == av.audio.frame.AudioFrame:
+                    audio_frames.append(frame.to_ndarray())
+                elif type(frame) == av.video.frame.VideoFrame:
+                    video_frames.append(frame.to_ndarray(format='rgb24'))
+
+                    if not only_keyframes and frame.key_frame == 1:
+                        curr_index = len(video_frames)
+                        keyframe_indices.append(curr_index)
+
+        video = parse_obj_as(VideoNdArray, np.stack(video_frames))
+
+        if only_keyframes:
+            return video
+        else:
+            if len(audio_frames) == 0:
+                audio = parse_obj_as(AudioNdArray, np.array(audio_frames))
+            else:
+                audio = parse_obj_as(AudioNdArray, np.stack(audio_frames))
+            indices = parse_obj_as(NdArray, keyframe_indices)
+            return audio, video, indices
```

Comment:
I am not a fan of having different return types depending on an argument. What about having a separate method for the `only_keyframes` case? What do you think @anna-charlotte @samsja?

Reply:
agree as well